go/enclave/storage/enclavedb/batch.go

```diff
@@ -0,0 +1,427 @@
+	selectHeader = "select b.header from batch b"
+
+	txExecInsert      = "insert into exec_tx values "
+	txExecInsertValue = "(?,?,?,?,?)"
+	queryReceipts     = "select exec_tx.receipt, tx.content, exec_tx.batch, batch.height from exec_tx join tx on tx.hash=exec_tx.tx join batch on batch.hash=exec_tx.batch "
+
+	selectTxQuery = "select tx.content, exec_tx.batch, batch.height, tx.idx from exec_tx join tx on tx.hash=exec_tx.tx join batch on batch.hash=exec_tx.batch where tx.hash=?"
+
+	selectContractCreationTx    = "select tx from exec_tx where created_contract_address=?"
+	selectTotalCreatedContracts = "select count( distinct created_contract_address) from exec_tx "
+
+	isCanonQuery = "select is_canonical from block where hash=?"
+)
+
+// WriteBatchAndTransactions - persists the batch and the transactions
+func WriteBatchAndTransactions(dbtx DBTransaction, batch *core.Batch) error {
+	bodyHash := batch.Header.TxHash.Bytes()
+
+	body, err := rlp.EncodeToBytes(batch.Transactions)
+	if err != nil {
+		return fmt.Errorf("could not encode L2 transactions. Cause: %w", err)
+	}
+	header, err := rlp.EncodeToBytes(batch.Header)
+	if err != nil {
+		return fmt.Errorf("could not encode batch header. Cause: %w", err)
+	}
+
+	dbtx.ExecuteSQL(bodyInsert, bodyHash, body)
+
+	var parentBytes []byte
+	if batch.Number().Uint64() > 0 {
+		parentBytes = batch.Header.ParentHash.Bytes()
+	}
+
+	// todo - this can be removed if the batches have no is_canonical
+	var isCanon bool
+	err = dbtx.GetDB().QueryRow(isCanonQuery, batch.Header.L1Proof.Bytes()).Scan(&isCanon)
+	if err != nil {
+		isCanon = false
+	}
+
+	dbtx.ExecuteSQL(bInsert,
+		batch.Hash().Bytes(),                   // hash
+		parentBytes,                            // parent
+		batch.Header.SequencerOrderNo.Uint64(), // sequence
+		batch.Header.Number.Uint64(),           // height
+		isCanon,                                // is_canonical
+		header,                                 // header blob
+		bodyHash,                               // reference to the batch body
+		batch.Header.L1Proof.Bytes(),           // l1_proof
+	)
+
+	// creates a big insert statement for all transactions
+	if len(batch.Transactions) > 0 {
+		insert := txInsert + strings.Repeat(txInsertValue+",", len(batch.Transactions))
+		args := make([]any, 0)
+		for i, transaction := range batch.Transactions {
+			txBytes, err := rlp.EncodeToBytes(transaction)
+			if err != nil {
+				return fmt.Errorf("failed to encode block receipts. Cause: %w", err)
+			}
+
+			args = append(args, transaction.Hash().Bytes()) // tx_hash
+			args = append(args, txBytes)                    // content
+			args = append(args, nil)                        // sender_address - todo - implement
+			args = append(args, transaction.Nonce())        // nonce
+			args = append(args, i)                          // idx
+			args = append(args, bodyHash)                   // the batch body which contained it
+		}
+		dbtx.ExecuteSQL(insert[0:len(insert)-1], args...)
```

Comment:
len() - 1 stuff was a bit confusing/distracting. Maybe worth doing it the line after the `Repeat` with a `// remove trailing comma comment` or something